Add blog post on GitHub Actions with GraalVM (#1098)
oktadeveloper/okta-blog: _source/_posts/2022-03-29-github-actions-graalvm.adoc

---
layout: blog_post
title: "Use GitHub Actions to Build GraalVM Native Images"
author: matt-raible
by: advocate
communities: [java,devops]
description: "Sometimes it takes a journey of many hours and many builds to produce a released artifact. GraalVM + GitHub Actions = ❤️."
tags: [graalvm,jhipster,springnative,jhipsternative]
tweets:
- ""
- ""
- ""
image:
type: conversion
---
:page-liquid:
:toc: macro
:experimental:

// keywords: github actions (74K), graalvm spring boot (590), github actions tutorial (2400), github ci cd (5400)

Getting something to work is one of the greatest feelings you can have as a developer. Especially when you've spent hours, days, or months trying to make it happen. The last mile can be one of the most painful and rewarding experiences, all wrapped into the same day or two.

> I experienced this recently with link:/blog/2022/03/03/spring-native-jhipster[Spring Native for JHipster]. If I look back, it took a year's worth of desire, research, and perseverance to make it happen. When we finally got it working&mdash;and automated&mdash;you can image my excitement!

Reviewer: ```suggestion
I experienced this recently with link:/blog/2022/03/03/spring-native-jhipster[Spring Native for JHipster]. If I look back, it took a year's worth of desire, research, and perseverance to make it happen. When we finally got it working&mdash;and automated&mdash;you can imagine my excitement!
```
typo

After a few days of euphoria, I thought it'd be easy to create native builds for each operating system (Linux, macOS, and Windows) using GitHub Actions. I was wrong.

If you'd like to follow along and learn how to configure GitHub Actions to create native binaries, you'll need a few prerequisites.

*Prerequisites*:

- https://sdkman.io[Java 11]+
- https://nodejs.org[Node 16]
- A https://github.com/signup[GitHub Account]

toc::[]

You can read the full conversation (aka a condensed version of this post) via the tweet below, or keep reading to walk through the trials and tribulations I experienced with GitHub Actions and GraalVM. Hopefully my learnings will save you hours of time.

++++
{% twitter 1496278460293324801 %}
++++

> == Configure a Full-Stack Native Java app to use GitHub Actions

Reviewer: ```suggestion
== Configure a full-stack native Java app to use GitHub Actions
```
quibble, quibble

Author: I also had a couple of other heading ideas in the comments below.

If you'd like a bit more background, please read https://auth0.com/blog/full-stack-java-with-react-spring-boot-and-jhipster/[Full Stack Java with React, Spring Boot, and JHipster] followed by link:/blog/2022/03/03/spring-native-jhipster[Introducing Spring Native for JHipster].
====

Clone the result of these two blog posts, right after I integrated the JHipster Native blueprint 1.0. Install its dependencies using npm.

[source,shell]
----
git clone -b jhipster-native-1.0 \
  https://github.com/oktadev/auth0-full-stack-java-example.git flickr2
cd flickr2
npm install
----

Then, create a https://github.com/new[new repo on GitHub]. For example, `jhipster-flickr2`.

Next, push this example project to it.

[source,shell]
----
USERNAME=<your-github-username>
git remote rm origin
git remote add origin git@github.com:${USERNAME}/jhipster-flickr2.git
git branch -M main
git push -u origin main
----

== Automate the wait with JHipster's CI/CD

Building native images with GraalVM brings me back to the days when we'd build Java apps with Ant and XDoclet back in the early 2000s. You'd start the build and go do something else for a while because it took _minutes_ for the artifacts to be built.

> There's another often-overlooked issue with native binaries: you have to build one for each operating system. It's not like Java where you can build a JAR and run it anywhere.

Reviewer: ```suggestion
There's another often-overlooked issue with native binaries: you have to build one for each operating system. It's not like Java where you can build a JAR (Java ARchive) and run it anywhere.
```
I suspect most of your readers will know this acronym, but I didn't...

Generate continuous integration (CI) configuration for GitHub Actions using https://www.jhipster.tech/setting-up-ci/[JHipster's CI/CD sub-generator].

[source,shell]
----

> jhipster ci-cd

Reviewer 2: Should add `npm install -g jhipster`? or recommend executing local jhipster `npx jhipster`?

Author: Good point. I'll fix this.

Author: Changed to `npx jhipster` since I do say to run `npm install` earlier.

----

This will prompt you to select a CI/CD pipeline. Select **GitHub Actions**.

image::{% asset_path 'blog/github-actions-graalvm/jhipster-ci-cd.png' %}[alt=Welcome to the JHipster CI/CD Sub-Generator,width=800, align=center]

> When prompted for the tasks/integrations (Sonar, Docker, Snyk, Heroku, and Cypress Dashboard), don't select any for this quick example. This will create a `.github/workflows/github-actions.yml` file. Rename it to `.github/workflows/main.yml` for simplicity.

Reviewer 3: This is an awesome feature!

Author: I recently [changed the default to `main.yml`](https://github.com/jhipster/generator-jhipster/pull/18101) in JHipster. I'll try and remember to update the post once the next version of JHipster is released.

> When prompted for the tasks/integrations (Sonar, Docker, Snyk, Heroku, and Cypress Dashboard), don't select any for this quick example. This will create a `.github/workflows/github-actions.yml` file. Rename it to `.github/workflows/main.yml` for simplicity.

Reviewer: ```suggestion
When prompted for the tasks/integrations for this quick example (Sonar, Docker, Snyk, Heroku, and Cypress Dashboard), don't select any. This will create a `.github/workflows/github-actions.yml` file. Rename it to `.github/workflows/main.yml` for simplicity.
```
reads better imo, because it puts some distance btwn "this" and "this"

[source,source]
----

> mv .github/workflows/github-actions.yml .github/workflows/main.yml

Reviewer 2: This won't be necessary at next jhipster version.

----

> This file configures GitHub Actions to checkout your project, configures Node 16, configures Java 11, run your project's backend tests, its frontend tests, and its end-to-end tests. Not only that, it'll start your dependent containers (e.g. Keycloak) in Docker. You can see that most of this functionality is hidden behind `npm run` commands.

Reviewer: ```suggestion
This file configures GitHub Actions to check out your project, configures Node 16, configures Java 11, and runs your project's backend tests, its frontend tests, and its end-to-end tests. Not only that, it'll start your dependent containers (e.g. Keycloak) in Docker. You can see that most of this functionality is hidden behind `npm run` commands.
```
check out is 2 words as a verb. 
confused by the verb forms here - is it GitHub Actions doing all the actions or the file that performs the other 'configure and run' tasks? Either way I think it should be "runs"

 Or maybe I've misunderstood?

Author: GitHub Actions does everything.

      - name: 'E2E: Package'
        run: npm run ci:e2e:package
      - name: 'E2E: Prepare'
        run: npm run ci:e2e:prepare
      - name: 'E2E: Run'
        run: npm run ci:e2e:run
        env:
          CYPRESS_ENABLE_RECORD: false
          CYPRESS_PROJECT_ID: ${{ secrets.CYPRESS_PROJECT_ID }}
          CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
      - name: 'E2E: Teardown'
        run: npm run ci:e2e:teardown
----
{% endraw %}

To test this out on your new repository, you'll need to create a branch and pull request (PR) with your changes.

[source,shell]
----
git checkout -b actions
git add .
git commit -m "Add GitHub Actions"
git push ci-cd actions
----

If you watch the tests run from your PR, you'll be pretty pleased until it hits the **E2E: Package** phase. It'll fail with the following error:

----
Error:  Failed to execute goal org.graalvm.buildtools:native-maven-plugin:0.9.10:build
  (build-native) on project flickr-2: Execution build-native of goal
  org.graalvm.buildtools:native-maven-plugin:0.9.10:build failed:
  GraalVM native-image is missing from your system.
Error:   Make sure that GRAALVM_HOME environment variable is present.

> ----

Reviewer 2: We may inject the GraalVM setup at the workflow.

Reviewer 2: With custom scripts, this will not fail anymore.

== The environmental impact of GraalVM builds

> This brings a fork in the road. Should you use the https://github.com/graalvm/setup-graalvm/issues/6[setup-graalvm action] to configure GraalVM and your Java SDK?

Reviewer: ```suggestion
This brings us to a fork in the road. Should you use the https://github.com/graalvm/setup-graalvm/issues/6[setup-graalvm action] to configure GraalVM and your Java SDK?
```
the idiom works better this way imo

> If you do, every time you create a PR, and commit to it, it will run a native build. These take a 3-4 minutes for me locally, but with GitHub Actions, it takes 30+ minutes! It'll also only build on Linux, so you don't know if there's issues with macOS or Windows.

Reviewer: ```suggestion
If you do, every time you create a PR, and commit to it, it will run a native build. These take a 3-4 minutes for me locally, but with GitHub Actions, it takes 30+ minutes! Also, it only builds on Linux, so you won't know if there are issues with macOS or Windows.
```
fussy grammar tweaks

> To me, this seems as bad for the environment as cryptocurrency. It'll also make you wish you bought crypto a couple of years ago if you're using a private repo. You only get 2000 free minutes of GitHub Actions for private repos. Any minutes after that, you get charged for.

Reviewer: ```suggestion
To me, this seems as bad for the environment as cryptocurrency. If you're using a private repo, it'll also make you wish you bought crypto a couple of years ago. You only get 2000 free minutes of GitHub Actions for private repos. Any minutes after that, you get charged for.
```
smoother read

> _Yes, I know the cryptocurrency topic is controversial. I do like to poke fun at it though. Native builds on every commit and mining bitcoin seem similar to me._

Author: Then again, simply [surfing the web](https://www.fastcompany.com/90717181/surfing-the-web-isnt-just-bad-for-your-brain-its-terrible-for-the-environment) is terrible for the environment too.

== Continuously test JVM builds with GitHub Actions

> There's a better way, in my opinion. Revert the changes made by the JHipster Native blueprint to the tasks that are run by the `npm run` commands. That is, change `ci:e2e:package` from `npm run native-package` to the following:

Reviewer 2: Due to required quirks that may be randomly required by native build, you could recommend a native executable daily build.
This could avoid a big bisect to find the problem.

Reviewer 2: And we may keep ci scripts untouched and change the blueprint workflow to use the custom scripts.

Author: > Due to required quirks that may be randomly required by native build, you could recommend a native executable daily build.

@mshima There is a sentence in the post:

> If you want to test things more often, you could do it with a [nightly build](https://kyuubang.github.io/devops/2022/01/22/nightly-build-with-github-actions.html).

As far as renaming `github-actions.yml` to `main.yml`, that should be in the next release (v7.7.1). I'll remove this part and test this tutorial again after it's released.

For the other things (e.g. keeping ci scripts untouched), I can pause the publishing of this post until after the next release of JHipster Native, or I can update this post (and the one it depends on) after it's released.

Reviewer 2: > If you want to test things more often, you could do it with a [nightly build](https://kyuubang.github.io/devops/2022/01/22/nightly-build-with-github-actions.html).

I saw, but it's in a different context, maybe move it here then.

You can do this quite easily with a GitHub Action that only runs when you create a release. If you want to test things more often, you could do it with a https://kyuubang.github.io/devops/2022/01/22/nightly-build-with-github-actions.html[nightly build].

=== Linux and Windows problems and solutions

When I first started trying to build native binaries with GraalVM, I quickly ran into https://github.com/graalvm/setup-graalvm/issues/6[issues] on Linux and Windows:

- Linux: `java.lang.OutOfMemoryError: GC overhead limit exceeded`
- Windows: `The command line is too long.`

I'm happy to say that I was able to fix the OOM error on Linux by specifying `-J-Xmx10g` in the build arguments of the `native-maven-plugin` plugin. Make this change in your project too.

[source,xml]
----
<plugin>
    <groupId>org.graalvm.buildtools</groupId>
    <artifactId>native-maven-plugin</artifactId>
    ..
    <configuration>
        <imageName>native-executable</imageName>
        <buildArgs>
            <buildArg>--no-fallback -J-Xmx10g</buildArg>
        </buildArgs>
    </configuration>
</plugin>
----

> TIP: If you to see more details from the `native-maven-plugin`, you can add `--verbose` as an additional argument.

Reviewer: ```suggestion
TIP: If you need to see more details from the `native-maven-plugin`, you can add `--verbose` as an additional argument.
```
typo - missing word

image::{% asset_path 'blog/github-actions-graalvm/release.png' %}[alt=Restore v0.0.1 - Giddyup!,width=800, align=center]

Click the **Actions** tab to watch your release execute. I'd like to warn you though, it's gonna take a while! My https://twitter.com/mraible/status/1498471457638293507[first successful release] took just under an hour.

- macOS: 31m 30s
- Linux: 33m 50s
- Windows: 59m 45s

However, I think you'll be pleased with the results. 🤠

image::{% asset_path 'blog/github-actions-graalvm/release-with-artifacts.png' %}[alt=Released with native binaries attached,width=800,align=center]

TIP: If your builds fail, you can delete the tag for the release by running `git push origin :v0.0.1`. Your release will then become a draft, and you can easily create the release again using the GitHub UI.

== Run your released binaries locally

If you were to download these binaries from GitHub and try to run them locally, you'd get failures because they can't connect to instances of Keycloak or PostgreSQL.

> To start up a PostgreSQL database for it to talk to, you can run the following command from your `flickr2` directory.

Reviewer: What's the "it" refer to in "for it to talk to" ? Unable to guess it.

Author: it == the JHipster app or your full-stack Java app, depending on how SEO-aggressive we want to be.

Author: @havi Do you think I should change "it" to something else?
